### PR TITLE
Limiting max block bytes

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: ref #7721

Problem Summary:

By default, the number of row of a block is `DEFAULT_BLOCK_SIZE`(2^16). If a query read a lot of columns, the reading bytes of each row can be large, for example 3KB, and a block can be about 200MB. This is extremely large and can cause OOM since there are many blocks processing concurrently.

### What is changed and how it works?

- Limiting max block bytes in query request. 
- The number of row of a block of background tasks, such as DeltaMerge, is `DEFAULT_MERGE_BLOCK_SIZE`, so we don't limit it. 
- Add configuration `profiles.default.max_block_bytes`, default value is `DEFAULT_BLOCK_BYTES`(16MB).
- `clipBlockRows` try to limit the block size not exceed `max_block_bytes` and it also make the clipped block rows align with `dt_segment_stable_pack_rows` since at least read one pack from DMFile each time.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 1]
```diff
@@ -2816,13 +2826,14 @@ BlockInputStreamPtr Segment::getLateMaterializationStream(BitmapFilterPtr && bit
     }
 
     // construct rest column stream
+    auto clipped_rest_block_rows = clipBlockRows(dm_context.db_context, expected_block_size, *rest_columns_to_read);
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:2829]: `clipped_rest_block_rows` should equal to `clipped_filter_block_rows`

File: dbms/src/Core/Defines.h [hunk 2]
```diff
@@ -64,6 +64,8 @@ static constexpr UInt64 DEFAULT_DISAGG_TASK_TIMEOUT_SEC = 5 * 60;
   */
 #define DEFAULT_BLOCK_SIZE 65536
 
+constexpr size_t DEFAULT_BLOCK_BYTES = DEFAULT_BLOCK_SIZE * 256; // 256B per row, total 16MB.
```
Comment (Lloyd-Pottiger) [inline dbms/src/Core/Defines.h:67]: maybe too small...

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 3]
```diff
@@ -2900,5 +2911,27 @@ BlockInputStreamPtr Segment::getBitmapFilterInputStream(const DMContext & dm_con
...
+        return expected_block_rows;
+    }
+    else
+    {
+        auto row_bytes = std::max(1, stable->avgRowBytes(read_columns)); // Avoid row_bytes to be 0.
+        auto rows = max_block_bytes / row_bytes;
+        rows = std::max(rows / pack_rows * pack_rows, pack_rows); // Align with pack and at least read one pack.
+        return std::min(expected_block_rows, rows);
+    }
+}
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:2934]: define as functions in anonymous namespace?

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 4]
```diff
@@ -2867,14 +2868,23 @@ BlockInputStreamPtr Segment::getBitmapFilterInputStream(const DMContext & dm_con
...
         return std::make_shared<EmptyBlockInputStream>(toEmptyBlock(columns_to_read));
     }
+
+    // buildBitmapFilter will only read handle, version and tag.
+    const ColumnDefines build_bitmap_columns{
+        getExtraHandleColumnDefine(is_common_handle),
+        getVersionColumnDefine(),
+        getTagColumnDefine(),
+    };
+    auto build_bitmap_block_rows = clipBlockRows(dm_context.db_context, expected_block_size, build_bitmap_columns);
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:2878]: There is no need to clip, just pass `expected_block_size` is ok.

File: dbms/src/Storages/DeltaMerge/StableValueSpace.cpp [hunk 5]
```diff
@@ -564,5 +564,53 @@ StableValueSpace::Snapshot::getAtLeastRowsAndBytes(const DMContext & context, co
...
+        return 8;
+    case Field::Types::UInt128:
+    case Field::Types::Int128:
+    case Field::Types::Decimal128:
+        return 16;
+    case Field::Types::Int256:
+    case Field::Types::Decimal256:
+        return 32;
+    case Field::Types::String:
+        return f.get<String>().size();
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/StableValueSpace.cpp:586]: Usually default value of string is `""` or `NULL`, which may cause inaccuracy.

Can we get `255` for type `varchar(255)`?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/StableValueSpace.cpp:586]: If the type is `varchar(255)`, default `""`, but the length of the data in this col is larger than `200`?

File: dbms/src/Storages/DeltaMerge/Segment.cpp (hunk 4 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:2878]: The memory usage of get bitmap is small, so I think pass `expected_block_size` is acceptable.

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:2878]: And those data will not pass to compute layer.

File: dbms/src/Core/Defines.h (hunk 2 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Core/Defines.h:67]: ok

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 6]
```diff
@@ -714,15 +714,17 @@ BlockInputStreamPtr Segment::getInputStream(const ReadMode & read_mode,
...
+        return getInputStreamModeNormal(dm_context, columns_to_read, segment_snap, read_ranges, filter ? filter->rs_operator : EMPTY_RS_OPERATOR, max_version, clipped_block_rows);
     case ReadMode::Fast:
-        return getInputStreamModeFast(dm_context, columns_to_read, segment_snap, read_ranges, filter ? filter->rs_operator : EMPTY_RS_OPERATOR, expected_block_size);
+        return getInputStreamModeFast(dm_context, columns_to_read, segment_snap, read_ranges, filter ? filter->rs_operator : EMPTY_RS_OPERATOR, clipped_block_rows);
     case ReadMode::Raw:
-        return getInputStreamModeRaw(dm_context, columns_to_read, segment_snap, read_ranges, expected_block_size);
+        return getInputStreamModeRaw(dm_context, columns_to_read, segment_snap, read_ranges, clipped_block_rows);
     case ReadMode::Bitmap:
+        // getBitmapFilterInputStream will read columns by multi-stages, so we clip its block rows inside.
         return getBitmapFilterInputStream(dm_context, columns_to_read, segment_snap, read_ranges, filter, max_version, expected_block_size);
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:728]:
```suggestion
        return getBitmapFilterInputStream(dm_context, columns_to_read, segment_snap, read_ranges, filter, max_version, expected_block_size, clipped_block_rows);
```

we can pass `expected_block_size` and  `clipped_block_rows` at the same time now.

**Review by Lloyd-Pottiger — APPROVED** [review]
Others LGTM

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 7]
```diff
@@ -2897,7 +2899,29 @@ BlockInputStreamPtr Segment::getBitmapFilterInputStream(const DMContext & dm_con
...
+{
+    if (unlikely(max_block_bytes <= 0)) // Disable block bytes limit.
+    {
+        return expected_block_rows;
+    }
+    else
+    {
+        auto row_bytes = std::max(1, stable->avgRowBytes(read_columns)); // Avoid row_bytes to be 0.
+        auto rows = max_block_bytes / row_bytes;
+        rows = std::max(rows / pack_rows * pack_rows, pack_rows); // Align with pack and at least read one pack.
```
Comment (SeaRise) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:2922]:
```suggestion
        rows = std::ceil(rows / pack_rows) * pack_rows; // Align with pack and at least read one pack.
```
just a little suggestion

**Review by SeaRise — APPROVED** [review]
Others LGTM

**Comment by SeaRise** [comment]
/hold
@JinheLin You can comment `/unhold` when you want to merge this pr.

**Pull request merged**